apps/emqx_bridge/src/emqx_action_info.erl

```diff
@@ -0,0 +1,197 @@
+    end.
+
+bridge_v1_type_to_action_type(Bin) when is_binary(Bin) ->
+    bridge_v1_type_to_action_type(binary_to_existing_atom(Bin));
+bridge_v1_type_to_action_type(Type) ->
+    ActionInfoMap = info_map(),
+    BridgeV1TypeToActionType = maps:get(bridge_v1_type_to_action_type, ActionInfoMap),
+    case maps:get(Type, BridgeV1TypeToActionType, undefined) of
+        undefined -> Type;
+        ActionType -> ActionType
+    end.
+
+action_type_to_bridge_v1_type(Bin) when is_binary(Bin) ->
+    action_type_to_bridge_v1_type(binary_to_existing_atom(Bin));
+action_type_to_bridge_v1_type(Type) ->
+    ActionInfoMap = info_map(),
+    ActionTypeToBridgeV1Type = maps:get(action_type_to_bridge_v1_type, ActionInfoMap),
+    case maps:get(Type, ActionTypeToBridgeV1Type, undefined) of
+        undefined -> Type;
+        BridgeV1Type -> BridgeV1Type
+    end.
+
+%% This function should return true for all inputs that are bridge V1 types for
+%% bridges that have been refactored to bridge V2s, and for all all bridge V2
+%% types. For everything else the function should return false.
+is_action_type(Bin) when is_binary(Bin) ->
+    is_action_type(binary_to_existing_atom(Bin));
+is_action_type(Type) ->
+    ActionInfoMap = info_map(),
+    ActionTypes = maps:get(action_type_names, ActionInfoMap),
+    case maps:get(Type, ActionTypes, undefined) of
+        undefined -> false;
+        _ -> true
+    end.
+
+registered_schema_modules() ->
+    InfoMap = info_map(),
+    Schemas = maps:get(action_type_to_schema_module, InfoMap),
+    maps:to_list(Schemas).
+
+%% ====================================================================
+%% Internal functions for building the info map and accessing it
+%% ====================================================================
+
+internal_emqx_action_persistent_term_info_key() ->
+    ?FUNCTION_NAME.
+
+info_map() ->
+    case persistent_term:get(internal_emqx_action_persistent_term_info_key(), not_found) of
+        not_found ->
+            build_cache();
+        ActionInfoMap ->
+            ActionInfoMap
+    end.
+
+build_cache() ->
+    ActionInfoModules = action_info_modules(),
+    ActionInfoMap =
+        lists:foldl(
+            fun(Module, InfoMapSoFar) ->
+                ModuleInfoMap = get_info_map(Module),
+                emqx_utils_maps:deep_merge(InfoMapSoFar, ModuleInfoMap)
+            end,
+            initial_info_map(),
+            ActionInfoModules
+        ),
+    %% Update the persistent term with the new info map
+    persistent_term:put(internal_emqx_action_persistent_term_info_key(), ActionInfoMap),
+    ActionInfoMap.
+
+action_info_modules() ->
+    ActionInfoModules = [
+        action_info_modules(App)
+     || {App, _, _} <- application:loaded_applications()
+    ],
+    lists:usort(lists:flatten(ActionInfoModules) ++ hard_coded_action_info_modules()).
+
+action_info_modules(App) ->
+    case application:get_env(App, emqx, action_info_module) of
```

Comment:
Taking Kafka app.src for example:

```erlang
{env, [{emqx_action_info_module, emqx_bridge_kafka_action_info}]},
```

Shouldn't this be:

```suggestion
    case application:get_env(App, emqx_action_info_module) of
```

? (also, `application:get_env/3` doesn't return `{ok, _}`)

Reply:
Good catch!